### PR TITLE
HPCC-11162 Improve spin block performance when contented

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln original pull request is https://github.com/hpcc-systems/HPCC-Platform/pull/5652

I'm not sure which is better, but both of them would reduce the number of yields when spin locks were contended.  Needs some testing for contended and uncontended cases - possibly by running the performance regression suite.

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -410,8 +410,23 @@ class jlib_decl  SpinLock
             nesting++;
             return;
         }
-        while (!atomic_cas(&value,1,0)) 
-            ThreadYield(); 
+
+        unsigned i = 0;
+        const unsigned maxSpins = 100;
```
Comment (mckellyln) [inline system/jlib/jmutex.hpp]: In standalone benchmark testing a value of 20 for maxSpins was better for all cases.

**Comment by mckellyln** [comment]
I recommend setting maxSpins to 10 to help the lightly contended cases and this does not affect the heavily contended cases.  Standalone micro benchmark testing shows no loss at 1,2,4 threads and large gains in performance at 8,16,24,32 threads (on 8 core machine) (with maxSpins set to 10)

**Comment by mckellyln** [comment]
while (!atomic_cas(&value,1,0))
    {
        while (atomic_read(&value))
        {
            if (i++ == maxSpins)
            {
                i = 0;
                thread_yield();
            }
        }
    }

performs slightly better, will look at asm to see if any sig difference.

**Comment by richardkchapman** [comment]
@ghalliday I think this is back in your court

**Comment by ghalliday** [comment]
I'll reimplement it on top of a fix for the spinblocks for power8/arm.

In particular I have added likely()/unlikely() macros which may help improve the code - they did for the spin block

**Comment by richardkchapman** [comment]
Typo in the title (contended vs contented)

**Comment by ghalliday** [comment]
Closing.  I think Mark should release a better version once the other spin pull request is merged (he can then call unlikely())

**Pull request closed without merging**